### ISSUE TITLE
Package source format

### ISSUE THREAD

**Issue by e9gille** [theaplroom/APM#2]
The code should be encapsulated in a top level namespace, but what form does it take in a package?

Options:
- Workspace file (dws)
Workspace file is certainly easy to go with and keeps everything in a single file, but it may cause issues with backwards compatibility (if trying to load a package in to an older Dyalog version).
- DCF file
Not sure about DCF files, in my experience storing namespaces in DCF files can suffer from similar version incompatibilities as with dws files.
- Single script file
A single class or namespace script is supported out of the box with Dyalog.
- Acre-Desktop
Would require Acre-Desktop of course.

**Comment by PaulMansour** [comment]
I'm generally for the workspace I think. 

Regarding a single script file, is this this practical?  Do you have any experience of a single script file containing dozens of potentially nested namespaces and thousands of functions?  Can this be created easily from a ws?   


But, I would note, there needs to be some software to load the package (unless it is a dyalog script)  , it could be Acre Desktop or it could be a separate package manager, but there needs to be  some software some where. If a dws is used to deliver the code, then it can contain the package managment code, and loading the workspace gets you access to the t 

The  paragraph immediately above is an unfinished thought.... please ignore for now.

**Comment by e9gille** [comment]
While you finish your thought, let me just say that I just listed the options I could think of. I don't like the idea of a single script file either. 

If APM is to work independently of acre, I think dws is the easiest way forward. It also makes it less likely that you end up modifying a package accidentally. 

I guess you could build the package using an appropriately old version of Dyalog to ensure compatibility.

**Comment by PaulMansour** [comment]
I think my main half finished point was that the dws delivery mechanism allows you to include the APM code for loading in the dependent projects.

However I'm wondering if it is a good idea to encourage the dws to be loaded directly by )load.  The dws is encapsulated in a folder which IS the package. In a way it is an implementation detail, and we should be able to change it .  If so, then the code to load a package the LoadPackage function, has to be part of the Interpreter, or Acre-Desktop or a user command library  ]APM if we want to keep it separate from Acre. 

Regardless it must be available during runtime to load package.

**Comment by PaulMansour** [comment]
I meant to add also that even if  APM is build into Acre, the dws is the best choice.  You could easily be delivering 50,000 files othewise.

**Comment by e9gille** [comment]
Yes, regardless of how a package is delivered I assume that a tool will load it into the workspace.
